### ISSUE TITLE
Sync.Once in client-go metrics play badly with components that want to provide them by default

### ISSUE THREAD

**Issue by alvaroaleman** [kubernetes/kubernetes#127739]
In Controller-Runtime, we register our metrics provider for the client-go [leaderelection](https://github.com/kubernetes-sigs/controller-runtime/blob/4381fa0aeee43e331be14b0d70cd276e1e91ad7a/pkg/metrics/leaderelection.go#L26), [workqueue](https://github.com/kubernetes-sigs/controller-runtime/blob/4381fa0aeee43e331be14b0d70cd276e1e91ad7a/pkg/metrics/workqueue.go#L99) and [clientmetrics](https://github.com/kubernetes-sigs/controller-runtime/blob/4381fa0aeee43e331be14b0d70cd276e1e91ad7a/pkg/metrics/client_go_adapter.go#L43-L54).

This is because controller-runtime provides a metrics endpoint and we want it to by default have all the metrics relevant to controllers.

Unfortunately, all this register funcs are internally a `sync.Once`. This means that if someone wants to register their own metrics, they have to do before someone else does. As of today, controller-runtime does this in an `init`, but even if it did that later, users would have to register this before controller-runtime does and make sure that no other dep registers it first, otherwise it will not work.

IMHO, we should remove the `sync.Once` so that as many metrics provides as wanted can be registered and this doesn't become a "first one wins, rest gets nothing" kind of situation where anyone who wants their custom adapter has to be super careful to be the first to register it to avoid it silently breaking.

There was some prior discussion around this specifically in the context of workqueue metrics, where its now possible to set a per-workqueue metrics provider that takes precedence over the global one as workaround:
* https://github.com/kubernetes/kubernetes/pull/114242

In that context, a PR to allow overriding the global one was rejetected, but i think we should be doing this: https://github.com/kubernetes/kubernetes/pull/116616

There was also some Slack discussion around this: https://kubernetes.slack.com/archives/C0EG7JC6T/p1719851021075269

This originally got reported as an issue in controller-runtime: https://github.com/kubernetes-sigs/controller-runtime/issues/2957

/sig api-machinery
/kind bug

**Comment by aojea** [comment]
/sig instrumentation
@logicalhan @dgrisonnet can you please chime in too?

**Comment by logicalhan** [comment]
We can probably just expose the register calls in a different file which is later invoked in a registration call with a sync.Once, that way you can get what you want and we don't have to change anything in k8s/k8s.

**Comment by alvaroaleman** [comment]
> We can probably just expose the register calls in a different file which is later invoked in a registration call with a sync.Once, that way you can get what you want and we don't have to change anything in k8s/k8s.

I don't understand how moving things around helps here, can you elaborate?

**Comment by dgrisonnet** [comment]
The main problem with how we register metrics is that we register all of them in a global registry via `init` functions, so we sometime can get noise inside the global metric registry, but I don't think this is a problem in your scenario.

You could create a handler serving both your registry and the Kubernetes global legacy registry and it should serve all the metrics that you need unless I am missing something. You can find an example of that here: https://github.com/kubernetes-sigs/metrics-server/blob/master/pkg/server/config.go#L105-L120

In the past I took a stab at moving away from the global registration system and the global registry and the result look like the following:
- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics/metrics.go#L66-L82 Introduces a new Register function where users can pass on their Registry for the metrics to be registered.
- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L319-L324 register in the application code

But it is difficult to do that globally as most of the codebase relies on the global registry and we would need to make sure that all the contributors follow the same pattern.

**Comment by alvaroaleman** [comment]
> You could create a handler serving both your registry and the Kubernetes global legacy registry and it should serve all the metrics that you need unless I am missing somethin

Right, rather than changing the global to not be a `sync.Once`, we could implement something in c-r that allows registering additional handlers. But that makes things strictly worse, because any downstream project needs to check if they have c-r in their dependency tree, if yes, register their handler in c-r, if not register it in client-go.

Can you please explain where the opposition to removing the `sync.Once` and allow registering multiple handlers come from? What issue do you see with that?

**Comment by dgrisonnet** [comment]
Allowing multiple handlers is a workaround from the Prometheus client, but I don't have the context as to why it was added in the first place, but it works well when you have to deal with multiple registries.

> Can you please explain where the opposition to removing the sync.Once

The fact that we need sync.Once in the first place is because most of the codebase uses the global legacy registry and it protects from scenarios when we would register the same metrics multiple times in the same registry.
The main push back I've seen is that we would break any consumer of the legacy registry that expects all metrics to be auto registered. I don't think anyone is heavily against cleaning up the sync.Once, but it would be a big and risky effort.

**Comment by alvaroaleman** [comment]
> The main push back I've seen is that we would break any consumer of the legacy registry that expects all metrics to be auto registered. I don't think anyone is heavily against cleaning up the sync.Once, but it would be a big and risky effort

Can you expand on the risk you are seeing? What risk is there to removing the sync.Once and replace it with a mutex and a slice of metrics provider? That way every provider that registers themselves will get invoked.

> An option that could be better for c-r would be to expose a Register function that would register all the c-r metrics to a registry provided by the caller. It would then be up to the consumer of c-r to create a handler that would serve both the metrics from c-r and the ones from client-go.

The problem here is that the majority of users of controller-runtime expect this to work by default without them having to do anything. If we change this, it will break for them and in a way where it does not break compilation and they will only notice at runtime that the metrics are not there. We want to retain this and have it work by default (majority case) while still making it possible to register custom handlers if client-go if wanted (minority/advanced case but still semi-frequently wanted).

**Comment by dgrisonnet** [comment]
> Can you expand on the risk you are seeing? What risk is there to removing the sync.Once and replace it with a mutex and a slice of metrics provider? That way every provider that registers themselves will get invoked.

_(note that the paragraph below and my previous comment is kind of unrelated to your initial problem and is more about the use of sync.Once for metrics in the codebase in general)_
The main problem is that consumer of client-go and other libraries in the codebase expect metrics to be registered by default in the legacyregistry. If we were to change to the cleaner mechanism of letting consumers register metrics themselves it'll break the expectations that consumers have today and have had for the past years and I'd expect most of them to end up in a situation where they don't collect the metrics anymore and won't be able to notice it until something bad happens.

The situation of https://github.com/kubernetes/client-go/blob/master/tools/metrics/metrics.go#L131-L170 is a bit different than the sync.Once situation that I am used to seeing in the codebase. Usually they are here to prevent registering the same metric twice and thus causing a panic. I would assume that this one is here mainly to avoid data races, but you also don't really want two different part of your code setting different providers as the result would depend on code ordering and might yield unexpected results for your users. In that case, a mutex wouldn't solve the issue.

> The problem here is that the majority of users of controller-runtime expect this to work by default without them having to do anything. If we change this, it will break for them and in a way where it does not break compilation and they will only notice at runtime that the metrics are not there. We want to retain this and have it work by default (majority case) while still making it possible to register custom handlers if client-go if wanted (minority/advanced case but still semi-frequently wanted).

Either way, having two different part of the code setting the provider is a problem. Depending on code ordering, it will either be set to the value from c-r or from its consumer. The former scenario where the last value is set by c-r will yield an unexpected result for your users.

I think the best option would be make it possible to opt-out from registering client-go providers in c-r when they run into the sync.Once issue. I don't really see how this could be addressed properly at the client-go level unless we want to be able to have multiple providers for the same metric factory which I don't know if it really makes a lot of sense.

--

Btw unrelated to the discussion above but I was checking the leader election code you shared and https://github.com/kubernetes-sigs/controller-runtime/blob/4381fa0aeee43e331be14b0d70cd276e1e91ad7a/pkg/metrics/leaderelection.go#L18-L21 is not registered so I don't think it will be exposed properly, unless I am missing something.

**Comment by dgrisonnet** [comment]
> unless we want to be able to have multiple providers for the same metric factory which I don't know if it really makes a lot of sense

Oh wait I missed the point my previous comment, but I guess that is what you are suggesting?

**Comment by alvaroaleman** [comment]
> Oh wait I missed the point in my previous comment, but I guess that is what you are suggesting?

Yeah. I do agree that "last registerer wins" would not really be any better than the current "first registerer wins". So allowing multiple providers to be registered is the best solution I can think of. WDYT?

**Comment by dgrisonnet** [comment]
Actually I just noticed that there is something I misunderstood from your user use-case. I though they wanted to import multiple libraries that were implementing the same metrics. But turns out they just want to extend what is provided by controller-runtime which make sense.

The three libraries you mentioned in the issue description have very different implementation so it might be hard to find a generic solution.
1. https://github.com/kubernetes/client-go/blob/master/util/workqueue/metrics.go has the notion of provider you mentioned which is an interface that allows to specify how to implement some metrics. You solution to allow multiple providers could make sense there but it leaves the door open to register the same interface multiple times and essentially producing duplicated data
2. https://github.com/kubernetes/client-go/blob/master/tools/metrics/metrics.go  has a struct with a bunch of metrics that users can provide. This one we could solve by adding sync.Once per metrics, although very ugly that would be effective. There might be other cleaner options though.
3. https://github.com/kubernetes/client-go/blob/master/tools/leaderelection/metrics.go has one single interface for essentially two metrics. For this one we could do the same as 1) but it as the same problem of allowing duplicated data.

**Comment by dgrisonnet** [comment]
It would be worth involving client-go maintainers. If they are fine with having potential duplicated data exposed by the clients, then your solution would simplify integration and extension and I don't think we have other options at least with the libraries that are using an interface.

**Comment by alvaroaleman** [comment]
> It would be worth involving client-go maintainers. If they are fine with having potential duplicated data exposed by the clients, then your solution would simplify integration and extension and I don't think we have other options at least with the libraries that are using an interface.

I do think that if there is code that might end up registering itself multiple times as provider, that code can and probably should deal with that by means of a sync.Once, rather than having the registry do that.

/cc @sttts @pohly 

How do you guys feel about the idea of allowing multiple metrics providers to be registered?

**Comment by dgrisonnet** [comment]
Looking back into this, I think we agreed that allowing multiple providers to be registered was the best option to fix this issue.

At this point, I think it should be fine to draft a PR for one of the impacted client-go library and see if the owners are in line with the new approach.

@alvaroaleman I don't have bandwidth to look into this myself, but if anyone wants to look into it feel free to step in.

**Comment by BitWeaverDev** [comment]
@JasonColapietro nice work on #141172. Before I start anything I'd rather check I'm not about to duplicate you: are you planning to carry the other two parts of this (`tools/metrics` and `workqueue`), or would it help if someone else picked one up?

If it's free I'd take `tools/metrics`. It looks like the part with the least room to work around. `workqueue` at least has the per-queue `MetricsProvider` config field, and informer metrics in `tools/cache` has a per-informer fallback, but `tools/metrics` has neither, so a second registrant there gets nothing and no escape hatch.

It's also shaped differently enough from leaderelection that I'd rather not guess at it in parallel with you. There's no per-instance construction point to snapshot at: the 16 hooks are exported package vars read straight off the request path (`rest/request.go`, `transport/cache.go`, the exec plugin), so fanning out means composing the vars in place, and keeping the single-provider case a plain direct assignment so the read path doesn't change at all. `RegisterFn`/`EnsureRegistered` needs the same treatment, and that part is easy to get wrong: the callback is deliberately lazy so registration lands after feature gates are set, so it can't just be fired eagerly from `Register`.

Two more `sync.Once` registrations in the same family that aren't listed on the issue, if it's useful to track them here:

- `tools/cache/informer_metrics.go`, `setInformerMetricsProviderOnce` (has a per-informer fallback)
- `tools/cache/reflector_metrics.go`, `metricsFactory.setProviders` (no override and no in-tree registrant, so it's currently dead weight either way)

Happy to wait either way, just don't want two people writing the same patch.

**Comment by JasonColapietro** [comment]
Thanks for checking and for digging into the edge cases. Please go ahead with `tools/metrics`; I’m not working on it. I’ll keep `workqueue` on my side for now, but I’ll comment here before starting so we don’t duplicate anything.
